### PR TITLE
Peggyv2 Solidity

### PR DESCRIPTION
**Pull request by jtremback**
This adds the `deployERC20` function to the contract which deploys an ERC20 whose entire balance is controlled by the Peggy.sol contract. It also tests it, as well as submitting a batch to transfer tokens out of it.

### CODE REVIEW

**Comment by jkilpatr** [comment]
looking good will review properly in the morning. Please ignore the validator out test as outlined in #160 it has issues that are being resolved.

File: solidity/contracts/CosmosToken.sol [hunk 1]
```diff
@@ -0,0 +1,16 @@
+pragma solidity ^0.6.6;
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+contract ComsosERC20 is ERC20 {
```
Comment (fedekunze) [inline solidity/contracts/CosmosToken.sol:4]:
```suggestion
contract CosmosERC20 is ERC20 {
```

File: solidity/contracts/Peggy.sol [hunk 2]
```diff
@@ -36,6 +37,15 @@ contract Peggy {
 		uint256 _amount,
 		uint256 _eventNonce
 	);
+	event ERC20DeployedEvent(
+		// FYI: Can't index on a string without doing a bunch of weird stuff
```
Comment (fedekunze) [inline solidity/contracts/Peggy.sol:41]: I don't understand this comment, can you elaborate?

File: solidity/contracts/Peggy.sol [hunk 3]
```diff
@@ -359,6 +363,27 @@ contract Peggy {
...
 	}
 
+	function deployERC20(
+		string memory _cosmosDenom,
+		string memory _name,
+		string memory _symbol,
+		uint8 _decimals
+	) public {
+		// Deploy an ERC20 with entire supply granted to Peggy.sol
+		ComsosERC20 erc20 = new ComsosERC20(address(this), _name, _symbol, _decimals);
```
Comment (fedekunze) [inline solidity/contracts/Peggy.sol:373]:
```suggestion
		CosmosERC20 erc20 = new CosmosERC20(address(this), _name, _symbol, _decimals);
```

File: solidity/contracts/Peggy.sol [hunk 4]
```diff
@@ -36,6 +37,15 @@ contract Peggy {
...
 		uint256 _eventNonce
 	);
+	event ERC20DeployedEvent(
+		// FYI: Can't index on a string without doing a bunch of weird stuff
+		string _cosmosDenom,
+		address _tokenContract,
+		string _name,
+		string _symbol,
+		uint8 _decimals,
+		uint256 _eventNonce
```
Comment (jkilpatr) [inline solidity/contracts/Peggy.sol:47]: we should index on the event nonce and tokencontract

**Review by jkilpatr — CHANGES_REQUESTED** [review]
Looks good we just need some minor changes to indexing and @fedekunze 's comment on the misspelled name handled.

File: solidity/contracts/CosmosToken.sol (hunk 1 above)
Comment (jtremback) [reply solidity/contracts/CosmosToken.sol:4]: lol thanks

File: solidity/contracts/Peggy.sol (hunk 2 above)
Comment (jtremback) [reply solidity/contracts/Peggy.sol:41]: We have an "indexed" argument on most of our events which makes them easier to look up. for some obscure reasons, strings cannot directly be indexed. You can make it work but i will wait until it's needed

File: solidity/contracts/Peggy.sol (hunk 4 above)
Comment (jtremback) [reply solidity/contracts/Peggy.sol:47]: Why are we indexing on any of this stuff?

**Comment by jtremback** [comment]
@jkilpatr @fedekunze addressed your review

**Review by jkilpatr — APPROVED** [review]

File: solidity/contracts/Peggy.sol (hunk 4 above)
Comment (jkilpatr) [reply solidity/contracts/Peggy.sol:47]: we don't use it but someone might. This is good now.

**Pull request merged**